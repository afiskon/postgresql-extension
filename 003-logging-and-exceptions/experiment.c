#include <postgres.h>
#include <utils/builtins.h>
#include <access/xact.h>

PG_MODULE_MAGIC;
PG_FUNCTION_INFO_V1(experiment_elog);
PG_FUNCTION_INFO_V1(experiment_ereport);
PG_FUNCTION_INFO_V1(experiment_trycatch);

Datum
experiment_elog(PG_FUNCTION_ARGS)
{
    /* int64 now = (int64)GetCurrentTransactionStartTimestamp(); */
    int64 now = 123456789;

    /* NB: %m corresponds to strerror(errno) */
    elog(NOTICE, "Transaction start timestamp: %lld",
        (long long)now);
    PG_RETURN_VOID();
}

Datum
experiment_ereport(PG_FUNCTION_ARGS)
{
    const char* arg = TextDatumGetCString(PG_GETARG_DATUM(0));
    ereport(ERROR,
        (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
         errmsg("Invalid argument \"%s\"", arg),
         errhint("This is a hint message.")));
    PG_RETURN_VOID();
}

Datum
experiment_trycatch(PG_FUNCTION_ARGS)
{
    PG_TRY();
    {
        elog(ERROR, "oops...");
    }
    PG_FINALLY();
    {
        elog(NOTICE, "cleaning up");
    }
    PG_END_TRY();

    PG_RETURN_VOID();
}