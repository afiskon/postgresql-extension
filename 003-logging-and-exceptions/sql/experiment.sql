CREATE EXTENSION experiment;

SELECT experiment_elog();

\set ON_ERROR_STOP 0
SELECT experiment_ereport('hello');
\set ON_ERROR_STOP 1

DROP EXTENSION experiment;