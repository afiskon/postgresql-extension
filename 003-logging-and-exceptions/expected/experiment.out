CREATE EXTENSION experiment;
SELECT experiment_elog();
NOTICE:  Transaction start timestamp: 123456789
 experiment_elog 
-----------------
 
(1 row)

\set ON_ERROR_STOP 0
SELECT experiment_ereport('hello');
ERROR:  Invalid argument "hello"
HINT:  This is a hint message.
\set ON_ERROR_STOP 1
DROP EXTENSION experiment;
